### PR TITLE
feat: human readable tfc workspace names

### PR DESCRIPTION
**Pull request by p0tr3c**
| Q                 | A
| ----------------- | ---
| 🐛 Bug fix?       | no
| 🚀 New feature?   | yes
| ⚠ Deprecations?   | no
| ❌ BC Break       | no
| 🔗 Related issues | #1251 
| ❓ Documentation  | [#197](https://github.com/snyk/driftctl-docs/pull/197)

## Description

Adds ability to resolve terraform cloud/enterprise human friendly names to corresponding workspace ID. 
The implementation attempts to discover the workspace ID via pattern matching in case existing users want to keep using the workspace id directly.

This PR also introduces the go-tfe client instead of using raw HTTP client. This could be controversial as it adds big module into the tools. I think this way we can reuse existing implementation of API and simplify the code. 

This was tested locally and appears to work for both human friendly names and workspace IDs. I've updated the tests and included test case for full path workspaces as well. 

This probably will trigger some conversations so posting it for early feedback. The change should include the documentation which I plan to add.

### CODE REVIEW

File: pkg/iac/terraform/state/backend/tfcloud_reader.go [hunk 1]
```diff
@@ -28,72 +31,93 @@ type TFCloudBody struct {
...
-		if res.StatusCode < 200 || res.StatusCode >= 400 {
-			return 0, errors.Errorf("error requesting terraform cloud backend state: status code: %d", res.StatusCode)
-		}
-
-		body := TFCloudBody{}
-		bodyBytes, _ := ioutil.ReadAll(res.Body)
-		err = json.Unmarshal(bodyBytes, &body)
+		stateVersion, err := t.client.StateVersions.Current(context.Background(), workspaceId)
 		if err != nil {
 			return 0, err
```
Comment (moadibfr) [inline pkg/iac/terraform/state/backend/tfcloud_reader.go:113]: Not sure what the error looks like, I was wondering if we should wrap it to give the user more context on what failed ?
like
```go
return 0,errors.WithMessage("error requesting terraform cloud backend state", err)
```

**Review by moadibfr — COMMENTED** [review]
Thanks for this first PR @p0tr3c :tada: 
I see little things to says apart from a small question on error handling.
It's cool that this can work with both ID and human readable name. I personally like that you choose to use the real tfe/tfc client that makes it more readable and should be more easy to maintain.

File: pkg/iac/terraform/state/backend/tfcloud_reader.go [hunk 2]
```diff
@@ -28,72 +31,93 @@ type TFCloudBody struct {
...
-			return err
+			return "", err
 		}
+		return reader.GetToken(u.Host)
+	}
+	return token, nil
+}
+
+// A regular expression used to validate string workspace ID patterns.
+var reStringID = regexp.MustCompile(`^ws-[a-zA-Z0-9\-\._]+$`)
```
Comment (eliecharra) [inline pkg/iac/terraform/state/backend/tfcloud_reader.go:82]: Are we sure of this regex ? I can't find any place where this is documented, how do you managed to build this one ?

**Review by eliecharra — CHANGES_REQUESTED** [review]
Thanks for this awesome PR @p0tr3c 🙏🏻 

> This PR also introduces the go-tfe client instead of using raw HTTP client. This could be controversial as it adds big module into the tools. I think this way we can reuse existing implementation of API and simplify the code.

I'm fine with that, if there is a library we should use it, I wasn't aware of this `go-tfe` thing 👍🏻

I posted a bunch of remarks, since we now use go-tfe we should also rework the way we are testing this part of the code.

File: pkg/iac/terraform/state/backend/tfcloud_reader.go (hunk 2 above)
Comment (p0tr3c) [reply pkg/iac/terraform/state/backend/tfcloud_reader.go:82]: Combination of [this](https://github.com/hashicorp/go-tfe/blob/main/workspace.go#L441) and following pattern that TFE has set in all workspaces. I don't think this is documented anywhere.

Comment (eliecharra) [reply pkg/iac/terraform/state/backend/tfcloud_reader.go:82]: I'm fine if this came from the tfe client 👍🏻 .

File: pkg/iac/terraform/state/backend/tfcloud_reader.go [hunk 3]
```diff
@@ -28,72 +31,93 @@ type TFCloudBody struct {
...
+	client        *tfe.Client
+	reader        io.ReadCloser
+	opts          *Options
+	workspacePath string
 }
 
-func NewTFCloudReader(client pkghttp.HTTPClient, workspaceId string, opts *Options) (*TFCloudBackend, error) {
-	req, err := http.NewRequest(http.MethodGet, fmt.Sprintf("%s/workspaces/%s/current-state-version", opts.TFCloudEndpoint, workspaceId), nil)
+func NewTFCloudReader(client pkghttp.HTTPClient, workspacePath string, opts *Options) (*TFCloudBackend, error) {
+	token, err := GetToken(opts)
```
[the remark this answers is not in the document]
Comment (p0tr3c) [reply pkg/iac/terraform/state/backend/tfcloud_reader.go:41]: I'll move the client initialization to read function. I am also going to remove the HTTPClient from the reader as it is not longer required.

File: pkg/iac/terraform/state/backend/tfcloud_reader_test.go [hunk 4]
```diff
@@ -27,20 +27,25 @@ func TestTFCloudBackend_Read(t *testing.T) {
...
-			url:      "https://app.terraform.io/api/v2/workspaces/workspaceId/current-state-version",
+			url:      "https://app.terraform.io/api/v2/workspaces/ws-ABCDEFG12345678/current-state-version",
 			wantErr:  nil,
 			expected: "{}",
 			mock: func() {
 				httpmock.Reset()
 				httpmock.RegisterResponder(
 					"GET",
-					"https://app.terraform.io/api/v2/workspaces/workspaceId/current-state-version",
+					"https://app.terraform.io/api/v2/ping",
```
[the remark this answers is not in the document]
Comment (p0tr3c) [reply pkg/iac/terraform/state/backend/tfcloud_reader_test.go]: Hope the changes address this comment 🤞

File: pkg/iac/terraform/state/backend/tfcloud_reader_test.go [hunk 5]
```diff
@@ -1,11 +1,14 @@
 package backend
 
 import (
-	"net/http"
+	"context"
 	"testing"
 
-	"github.com/jarcoal/httpmock"
+	"github.com/cloudskiff/driftctl/test/mocks"
```
Comment (eliecharra) [inline pkg/iac/terraform/state/backend/tfcloud_reader_test.go]: `s/cloudskiff/snyk/` everywhere 😉

File: pkg/iac/terraform/state/backend/tfcloud_reader_test.go [hunk 6]
```diff
@@ -1,11 +1,14 @@
...
 	"github.com/pkg/errors"
+
+	tfe "github.com/hashicorp/go-tfe"
+	"github.com/jarcoal/httpmock"
```
Comment (eliecharra) [inline pkg/iac/terraform/state/backend/tfcloud_reader_test.go:10]: This one should be removed

**Review by eliecharra — CHANGES_REQUESTED** [review]

File: pkg/iac/terraform/state/backend/tfcloud_reader_test.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (eliecharra) [reply pkg/iac/terraform/state/backend/tfcloud_reader_test.go]: Yeah thanks this is a way better 🙏🏻

File: pkg/iac/terraform/state/backend/tfcloud_reader.go [hunk 7]
```diff
@@ -28,72 +28,103 @@ type TFCloudBody struct {
...
+	return v != "" && reStringID.MatchString(v)
+}
+
+func (t *TFCloudBackend) GetWorkspaceId() (string, error) {
+	if isValidWorkspaceID(t.workspacePath) {
+		return t.workspacePath, nil
+	}
+	workspacePath := strings.Split(t.workspacePath, "/")
+	if len(workspacePath) != 2 {
+		return "", errors.New("unable to parse terraform-workspace, it should be either a workspace id (ws-xxxxx) or a {org}/{workspaceName}")
```
[the remark this answers is not in the document]
Comment (p0tr3c) [reply pkg/iac/terraform/state/backend/tfcloud_reader.go:79]: hahaha this took me a while to process :-D

File: pkg/iac/terraform/state/backend/tfcloud_reader.go [hunk 8]
```diff
@@ -28,72 +28,103 @@ type TFCloudBody struct {
...
 	}
-	t.request.Header.Add("Authorization", fmt.Sprintf("Bearer %s", token))
+	workspace, err := t.client.Workspaces.Read(context.Background(), workspacePath[0], workspacePath[1])
+	if err != nil {
+		return "", errors.Errorf("unable to read terraform workspace id: %s", err.Error())
+	}
+	return workspace.ID, nil
+}
+
+func (t *TFCloudBackend) InitTFEClient() error {
```
[the remark this answers is not in the document]
Comment (p0tr3c) [reply pkg/iac/terraform/state/backend/tfcloud_reader.go:88]: I agree it does not need to be exported, but not sure I follow the last line. In order to satisfy interface I thought you only are required to implement the specific functions. Having more functions does not impact the satisfaction criteria right?

[the remark this answers is not in the document]
Comment (eliecharra) [reply pkg/iac/terraform/state/backend/tfcloud_reader.go]: Right, but here we do not want to have something else exported

File: pkg/iac/terraform/state/backend/tfcloud_reader.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (eliecharra) [reply pkg/iac/terraform/state/backend/tfcloud_reader.go]: haha sorry we were sed-ing ourselves before but this is not obvious indeed 😅

[the remark this answers is not in the document]
Comment (p0tr3c) [reply pkg/iac/terraform/state/backend/tfcloud_reader.go]: I like it :-D

File: pkg/iac/terraform/state/backend/tfcloud_reader.go [hunk 9]
```diff
@@ -28,72 +28,103 @@ type TFCloudBody struct {
...
-	}
-	req.Header.Add("Content-Type", "application/vnd.api+json")
-	return &TFCloudBackend{req, client, nil, opts}, nil
+func NewTFCloudReader(workspacePath string, opts *Options) (*TFCloudBackend, error) {
+	return &TFCloudBackend{nil, nil, opts, workspacePath}, nil
 }
 
-func (t *TFCloudBackend) authorize() error {
-	token := t.opts.TFCloudToken
+func GetToken(opts *Options) (string, error) {
```
[the remark this answers is not in the document]
Comment (eliecharra) [reply pkg/iac/terraform/state/backend/tfcloud_reader.go]: Please set the pointer recieved `(t *TFCloudBackend)` we want to scope this `getToken` func to the struct, it does not make sense to expose it to the package.

**Review by eliecharra — APPROVED** [review]
That looks great 🎉 
Thanks @p0tr3c this one will be very useful for everyone for sure 👍🏻

**Comment by eliecharra** [comment]
@p0tr3c Can you please add a documentation PR for that, thanks 🙏🏻

**Comment by p0tr3c** [comment]
Needed to squash all the changes before merging.

**Review by eliecharra — APPROVED** [review]

**Comment by eliecharra** [comment]
@all-contributors please add @p0tr3c for code

**Comment by eliecharra** [comment]
@all-contributors please add @p0tr3c for code

**Pull request merged**